org.osgi.test.junit5.cm/src/test/java/org/osgi/test/junit5/cm/ConfigAnnotationTest.java

```diff
@@ -0,0 +1,146 @@
+package org.osgi.test.junit5.cm;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertNull;
+
+import org.junit.jupiter.api.AfterAll;
+import org.junit.jupiter.api.BeforeAll;
+import org.junit.jupiter.api.Nested;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.ExtendWith;
+import org.osgi.framework.BundleContext;
+import org.osgi.framework.FrameworkUtil;
+import org.osgi.framework.ServiceReference;
+import org.osgi.service.cm.Configuration;
+import org.osgi.service.cm.ConfigurationAdmin;
+import org.osgi.test.common.annotation.config.ConfigEntry;
+import org.osgi.test.common.annotation.config.InjectConfiguration;
+
+@ExtendWith(ConfigurationExtension.class)
+@Nested
+public class ConfigAnnotationTest {
+
+	public static final String							FACTORY_CONFIGURATION_PID	= "my.factory.configuration.pid";
+	public static final String							NONSTATIC_CONFIGURATION_PID	= "nonstatic.configuration.pid";
+
+	private static ServiceReference<ConfigurationAdmin>	sref						= null;
+	private static BundleContext						bundleContext				= null;
+	private static ConfigurationAdmin					ca							= null;
+
+	@BeforeAll
+	public static void beforeEach() {
+
+		bundleContext = FrameworkUtil.getBundle(ConfigAnnotationTest.class)
+			.getBundleContext();
+
+		sref = bundleContext.getServiceReference(ConfigurationAdmin.class);
+		if (sref == null) {
+			throw new IllegalStateException("The configuration admin service cannot be found.");
+		}
+
+		ca = bundleContext.getService(sref);
+		if (ca == null) {
+			throw new IllegalStateException("The configuration admin service cannot be found.");
+		}
+
+	}
+
+	@AfterAll
+	public static void afterEach() {
+		BundleContext bundleContext = FrameworkUtil.getBundle(ConfigAnnotationTest.class)
+			.getBundleContext();
+		if (bundleContext != null && sref != null) {
+			bundleContext.ungetService(sref);
+		}
+
+	}
+
+	// START TESTS
+	@InjectConfiguration(pid = NONSTATIC_CONFIGURATION_PID)
+	Configuration nonStaticConfiguration;
+
+	@Test
+	public void testFieldConfiguration() throws Exception {
+
+		Configuration cs = ConfigAdminUtil.getConfigsByServicePid(ca, NONSTATIC_CONFIGURATION_PID);
+
+		assertThat(cs).isEqualTo(nonStaticConfiguration);
+
+	}
+
+	static final String PARAM_PID = "param.pid";
+
+	@Test
+	public void testParameterConfiguration(@InjectConfiguration(pid = PARAM_PID, properties = {
+		@ConfigEntry(key = "bar", value = "foo")
+	}) Configuration configuration) throws Exception {
+
+		Configuration cs = ConfigAdminUtil.getConfigsByServicePid(ca, PARAM_PID);
+		assertThat(cs).isEqualTo(configuration);
```

Comment:
Should we perhaps assert here `isSameAs()` rather than merely `isEqualTo()`? Or do you envisage a valid use case where the injected configuration might be a different instance to the one returned by the `ConfigAdminUtil.getConfigsByServicePid()`?

Reply:
isSame is correct

Reply:
Surprise, we have to use Equals.

Felix ConfigurationAdminImpl creates a new ConfigurationAdapter each call of listConfigurations
```
    private Configuration wrap( ConfigurationImpl configuration )
    {
        return new ConfigurationAdapter( this, configuration );
    }
```